airflow/models/dag_version.py

```diff
@@ -0,0 +1,155 @@
+    serialized_dag = relationship(
+        "SerializedDagModel",
+        back_populates="dag_version",
+        uselist=False,
+        cascade="all, delete, delete-orphan",
+        cascade_backrefs=False,
+    )
+    dag_runs = relationship("DagRun", back_populates="dag_version", cascade="all, delete, delete-orphan")
+    task_instances = relationship("TaskInstance", back_populates="dag_version")
+    created_at = Column(UtcDateTime, default=timezone.utcnow)
+
+    __table_args__ = (
+        UniqueConstraint("dag_id", "version_number", name="dag_id_v_name_v_number_unique_constraint"),
+    )
+
+    def __repr__(self):
+        return f"<DagVersion {self.dag_id} {self.version}>"
+
+    @classmethod
+    @provide_session
+    def write_dag(
+        cls,
+        *,
+        dag_id: str,
+        version_name: str | None = None,
+        version_number: int = 1,
+        session: Session = NEW_SESSION,
+    ) -> DagVersion:
+        """Write a new DagVersion into database."""
+        existing_dag_version = session.scalar(
+            with_row_locks(cls._latest_version_select(dag_id), of=DagVersion, session=session, nowait=True)
+        )
+        if existing_dag_version:
+            version_number = existing_dag_version.version_number + 1
+
+        dag_version = DagVersion(
+            dag_id=dag_id,
+            version_number=version_number,
+            version_name=version_name,
+        )
+        log.debug("Writing DagVersion %s to the DB", dag_version)
+        session.add(dag_version)
+        # Flush is necessary here due to the unique constraint and other linked tables
+        session.flush()
+        log.debug("DagVersion %s written to the DB", dag_version)
+        return dag_version
+
+    @classmethod
+    def _latest_version_select(cls, dag_id: str) -> Select:
+        return select(cls).where(cls.dag_id == dag_id).order_by(cls.created_at.desc()).limit(1)
+
+    @classmethod
+    @provide_session
+    def get_latest_version(cls, dag_id: str, session: Session = NEW_SESSION) -> DagVersion | None:
+        return session.scalar(cls._latest_version_select(dag_id))
+
+    @classmethod
+    @provide_session
+    def get_version(
+        cls,
+        dag_id: str,
+        version_number: int = 1,
+        session: Session = NEW_SESSION,
+    ) -> DagVersion | None:
+        version_select_obj = (
+            select(cls)
+            .where(cls.dag_id == dag_id, cls.version_number == version_number)
+            .order_by(cls.version_number.desc())
+            .limit(1)
+        )
+        return session.scalar(version_select_obj)
+
+    @property
+    def version(self) -> str:
+        """Return the version name."""
+        name = f"{self.version_number}"
+        if self.version_name:
+            name = f"{self.version_name}-{self.version_number}"
+        return name
+
+    @classmethod
+    @provide_session
+    def get_latest_dag_versions(cls, dag_ids: list[str], session: Session = NEW_SESSION) -> list[DagVersion]:
```

Comment:
It seems weird this returns a list of versions instead of a dag_id -> version mapping.

Reply:
Since the DagVersion has the dag_id column, is it necessary to do a mapping?

Reply:
No but I guess the question is why do we need this function in the first place. I searched and it seems the only place this is called is so we can get a list of serialized DAGs, so maybe we should just have a function for this specific purpose instead. Or just fetch the serialized DAGs where they are needed (in `dags_needing_dagruns`) directly and just remove the additional layer entirely.

Reply:
I moved the method to the Serdag side